### ISSUE TITLE
Advice for porting custom swiper function to consult

### ISSUE THREAD

**Issue by dalanicolai** [minad/consult#625]
Hey Daniel M!

I am working on another imrproved display engine for Emacs (just for 'motivation', here is a screenshot)
![image](https://user-images.githubusercontent.com/18429791/183768775-ae267a8e-1533-4895-840b-dca1128d4d4f.png)

It can show documents in as many columns as desired, and also it offers a thumbnail side window, which again can be formatted in as many columns as desired (all working perfectly, fast and smoothly). These are some nice improvement on the [image-roll](https://github.com/dalanicolai/image-roll.el) package).

Anyway, another cool feature I have been working on, because it is much more powerful while much easier to implement than `isearch`,
is swiper like functionality, to search through documents. The nice thing is, that it does not just work on lines, but on any 'type' of text block that is defined in the document source, again here is a short video for motivation
[Screencast from 07-26-2022 05:09:51 PM.webm](https://user-images.githubusercontent.com/18429791/183764517-6bbcb01c-39a4-4e13-9883-58d68fc3c250.webm)

As you can see in the video, it already works for searching over full paragraphs (and it be configured to 'swipe' by page, line or word, simply by setting a variable).

However, I have implemented the functionality as shown in the video using de ivy/swiper framework. Now I was trying to port it to consult/vertico, as consult is compatible with all completion engines. However, a problem that I encountered is that, when I am feeding the complete paragraphs to `consult--read`, consult--read does not format them nicely (it removes the newline characters), and it also modifies the candidates in some way that I can not find results that should be there (I do find them when I print the list of candidates directly to the message buffer).

Do you think it would be possible for consult to support the functionality as shown in the video? That would be great!

Finally, if you think it will be possible, then maybe you also have some suggestions for how you think this could best be implemented.
I guess I would have to define another `:category` as getting markers for each 'text-block- is probably not the best way to implement it.
Well, let's first await your answer about a possible solution for the 'candidates' issue first...

**Comment by minad** [comment]
The newlines in candidates are removed by vertico--truncate-multiline, since Vertico assumes single line candidates. We could make this formatting step in Vertico optional such that you can display multiline candidates unmodified. For now you can replace the function with an advice to disable it.

It should be possible to port the Ivy functionality directly to Consult, e.g. the automatic jumping and highlighting. See for example consult-line.

**Comment by dalanicolai** [comment]
Will have a look. I was taking consult-line as an example already, for finding out how consult works. Thanks for the quick reply!

**Comment by minad** [comment]
I close this issue. Feel free to continue the discussion anyway. Did you make some progress?

**Comment by dalanicolai** [comment]
That's fine. Well, I worked on it until I thought it should be working, but, do to some unexpected text structure i.c.w. some unexpected behavior, it seemed to not work yet. So then I thought I'll give it a break, and come back to it later, then worked on other priorities. But now that you're asking, I quickly went back, and found it was already implemented just fine the last time. Today, I checked the structured text in the file, and concluded that things were working alright, but I just had to list things unsorted, and use orderless to get the orderless search behavior. So indeed everything is working just fine, although for the final polish, I have to look another time at how to adapt the `consult--jump-state`, but I guess that would not be too much work also. Thanks, for asking. Also, now I finally understand, from inspection, why Vertico and orderless are great (I have looked now at the design difference and finally understand it a bit, using the advice instead of the `completing-read-fuction` variable, and design consult to work with the `completing-read-default' function). Great work!

**Comment by dalanicolai** [comment]
Ah, I was a bit too quick, I now see that I was still using Ivy. Anyway, Vertico indeed list things just fine after commenting out
```
                  (when (string-match-p "\n" cand)
                    (setq cand (vertico--truncate-multiline cand max-width)))
```
in `vertico--arrange-candidates`. Would be nice to make that behavior configurable depending on the 'calling-function' or, maybe for consult-read, on 'category', but I have not looked into how `category` works yet (e.g. how to define a new one).

**Comment by minad** [comment]
> Would be nice to make that behavior configurable depending on the 'calling-function' or, maybe for consult-read, on 'category', but I have not looked into how category works yet (e.g. how to define a new one).

We could make that configurable, however right now Vertico relies on candidates fitting on a single line. Changing this would also require changes to the resizing code etc. That's the reason this hasn't been done yet and also to avoid complexity.

**Comment by dalanicolai** [comment]
I am not sure what you mean with 'relies on candidate fitting on a single line', as with those two lines filtered out, things currently seems to work perfectly fine. Maybe you mean this would not support grouping or narrowing, or dispatching in Embark? Anyway, those functionalities are not needed here and could be excluded for the 'multiple line candidates'.

But anyway, that is not why I came here. Because I have a(nother) question. As you can see in the screencast I in my first comment, I am not only highlighting the current result in the document, but I am also highlighting the other matches (the remaining candidates left after filtering). Here is a screenshot (from that video) that shows the functionality
![image](https://user-images.githubusercontent.com/18429791/186069642-38c213e6-feb4-47ca-91d4-1c0345b843c2.png)
 You can see the current candidate is highlighted with 'yellow' background, which thus also indicates the search 'unit' (unit, is a paragraph, but we could set it to page, or line to get the equivalent of 'consult-line'). To be able to highlight the other 'remaining candidates (the other matches in the screenshot, without the yellow background), in Ivy I am using the variable called 'filtered-candidates', which gives a list of the 'remaining' (i.e. filtered) candidates in the minibuffer, but I am not able to quickly find the consult equivalent of this variable. It looks like the `lookup` functions only receives `selected`, `(all) candidates`, and the `input` (and optionally a few more things), but not the filtered candidates.

Do you have a suggestion for how I could easiest obtain these 'filtered candidates'?

**Comment by minad** [comment]
> I am not sure what you mean with 'relies on candidate fitting on a single line', as with those two lines filtered out, things currently seems to work perfectly fine.

Mostly minibuffer resizing won't work well.

> Do you have a suggestion for how I could easiest obtain these 'filtered candidates'?

Consult or the UI doesn't allow accessing the other filtered candidates. In Vertico the candidates are available as vertico--candidates, but at this point you lose the UI independence.

You could also use the completion table and manually filter again using completion-all-completions.

**Comment by minad** [comment]
In order to support this we would need some abstraction layer to obtain all candidates from the UI. So far this hasn't been needed, which leads to very little integration code (e.g. consult-vertico.el). 

The point here is that we are approaching the problem differently. When I designed Consult, I wanted to see how much I can achieve with minimal means. Your approach is different - you put out a fixed set of goals and then look for a possible implementation. This necessarily leads to stronger coupling of components.

You could either implement one of the suggestions from above (special casing Vertico etc, double filtering via completion-all-completions, implement a separate abstraction to access all candidates). But I wonder, do we really need to see all matches highlighted in the PDF too? Did you notice that I also don't do this for consult-line? So far I am happy with that since I also see all the matches in the minibuffer. 

Having less clutter is also not bad. I've also gotten surprisingly few complaints about this. It could very well be that users who want to replicate the Ivy behavior precisely just turn away after having tried Consult for a short while. But then I've also not gotten reports like "I would use Consult if not for this or that particular feature".

Anyway, there is nothing wrong with using Ivy. I never intended to make this an exact replacement. It all boils down to preference. I wanted to have simpler, smaller, more general and more modular components. Otoh for me it is okay to compromise here or there.

**Comment by dalanicolai** [comment]
I understand your approach, and I like it. I am actually doing a similar thing for my 'doc-view alternative'. Your point about the necessity to see all matches, is valid indeed. Initially, I intended the swiper functionality to replace the isearch functionality, as isearch support in the doc-viewer is a pain to implement. For isearch, I do like the highlighting of other matches. Also, while for consult-line the other candidates are formatted like in the buffer, this is not the case for the document consult-line alternative.
But anyway, I have implemented a simple word search now next to the swiper functionality, so that for swiper the highlighting of other matches is (even) less important. So I guess I will probably just provide support for both consult and swiper and provide them as separate extension packages.
